File: 2015/day07_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from functools import lru_cache
from typing import Any, Literal

type Instruction = tuple[
    str,
    str,
    Literal["SIG"]
    | Literal["NOT"]
    | Literal["AND"]
    | Literal["OR"]
    | Literal["LSHIFT"]
    | Literal["RSHIFT"],
    str,
]


instructions: list[Instruction] = []


def solution(content: list[str]) -> Any:
    # Using global statement to enable cache on `evaluate`
    global instructions  # pylint: disable=global-statement
    instructions = [parse(line) for line in content]
    return evaluate(target("a"))


@lru_cache
def evaluate(i: Instruction):
    left, right, op, _ = i
    if op == "SIG" and is_number(left):
        return int(left)
    if op == "SIG":
        return evaluate(target(left))
    if op == "NOT" and is_number(left):
        return norm(~int(left))
    if op == "NOT":
        return norm(~evaluate(target(left)))
    if op == "AND" and is_number(left) and is_number(right):
        return int(left) & int(right)
    if op == "AND" and is_number(left):
        return int(left) & evaluate(target(right))
    if op == "AND" and is_number(right):
        return int(right) & evaluate(target(left))
    if op == "AND":
        return evaluate(target(left)) & evaluate(target(right))
    if op == "OR" and is_number(left) and is_number(right):
        return int(left) | int(right)
    if op == "OR" and is_number(left):
        return int(left) | evaluate(target(right))
    if op == "OR" and is_number(right):
        return int(right) | evaluate(target(left))
    if op == "OR":
        return evaluate(target(left)) | evaluate(target(right))
    if op == "LSHIFT" and is_number(left):
        return int(left) << int(right)
    if op == "LSHIFT":
        return evaluate(target(left)) << int(right)
    if op == "RSHIFT" and is_number(left):
        return int(left) >> int(right)
    if op == "RSHIFT":
        return evaluate(target(left)) >> int(right)
    raise ValueError(f"Invalid operation {op}")


def norm(val: int):
    if val < 0:
        return 65536 + val
    return val


def target_of(i: Instruction):
    return i[3]


def target(trg: str) -> Instruction:
    return [t for t in instructions if target_of(t) == trg][0]


def parse(line: str) -> Instruction:
    parts = line.strip().split(" ")
    if "NOT" in parts:
        return (parts[1], "", "NOT", parts[-1])
    if "AND" in parts:
        return (parts[0], parts[2], "AND", parts[-1])
    if "OR" in parts:
        return (parts[0], parts[2], "OR", parts[-1])
    if "LSHIFT" in parts:
        return (parts[0], parts[2], "LSHIFT", parts[-1])
    if "RSHIFT" in parts:
        return (parts[0], parts[2], "RSHIFT", parts[-1])
    return (parts[0], "", "SIG", parts[-1])


def is_number(val: str):
    return all(ch in "1234567890" for ch in val)


def main() -> None:
    with open("day07.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()

File: 2015/day07-1.hs
import Data.Bits (Bits (complement, shiftL, shiftR, (.&.), (.|.)))
import Data.Map (Map, empty, insert, lookup, member)
import Data.Maybe (fromJust, fromMaybe)
import Prelude hiding (lookup)

data Operator = SIG | NOT | AND | OR | LSHIFT | RSHIFT deriving (Show)

type Instruction = (String, String, Operator, String)

type Cache = Map String Int

solution :: String -> Int
solution = evaluate "a" empty . map parse . lines

evaluate :: String -> Cache -> [Instruction] -> Int
evaluate key cache ins
  | member key cache = fromJust . lookup key $ cache
  | otherwise = evaluate key newCache toDo
  where
    toEval = filter (isEvaluatable cache) ins
    toDo = filter (not . isEvaluatable cache) ins
    newCache = foldl doEval cache toEval

doEval :: Cache -> Instruction -> Cache
doEval cache (left, _, SIG, t) = insert t (fromMaybe (read left) (lookup left cache)) cache
doEval cache (left, _, NOT, t) = insert t (complement $ fromMaybe (read left) (lookup left cache)) cache
doEval cache (left, right, AND, t) = insert t ((.&.) (fromMaybe (read left) (lookup left cache)) (fromMaybe (read right) (lookup right cache))) cache
doEval cache (left, right, OR, t) = insert t ((.|.) (fromMaybe (read left) (lookup left cache)) (fromMaybe (read right) (lookup right cache))) cache
doEval cache (left, right, LSHIFT, t) = insert t (shiftL (fromMaybe (read left) (lookup left cache)) (fromMaybe (read right) (lookup right cache))) cache
doEval cache (left, right, RSHIFT, t) = insert t (shiftR (fromMaybe (read left) (lookup left cache)) (fromMaybe (read right) (lookup right cache))) cache

isEvaluatable :: Cache -> Instruction -> Bool
isEvaluatable cache (left, _, SIG, _) = isNumber left || member left cache
isEvaluatable cache (left, _, NOT, _) = isNumber left || member left cache
isEvaluatable cache (left, right, AND, _) = (isNumber left || member left cache) && (isNumber right || member right cache)
isEvaluatable cache (left, right, OR, _) = (isNumber left || member left cache) && (isNumber right || member right cache)
isEvaluatable cache (left, right, LSHIFT, _) = (isNumber left || member left cache) && (isNumber right || member right cache)
isEvaluatable cache (left, right, RSHIFT, _) = (isNumber left || member left cache) && (isNumber right || member right cache)

isNumber :: String -> Bool
isNumber = all (`elem` "1234567890")

parse :: String -> Instruction
parse line
  | (!! 0) ws == "NOT" = ((!! 1) ws, "", NOT, (!! 3) ws)
  | (!! 1) ws == "AND" = ((!! 0) ws, (!! 2) ws, AND, (!! 4) ws)
  | (!! 1) ws == "OR" = ((!! 0) ws, (!! 2) ws, OR, (!! 4) ws)
  | (!! 1) ws == "LSHIFT" = ((!! 0) ws, (!! 2) ws, LSHIFT, (!! 4) ws)
  | (!! 1) ws == "RSHIFT" = ((!! 0) ws, (!! 2) ws, RSHIFT, (!! 4) ws)
  | otherwise = ((!! 0) ws, "", SIG, (!! 2) ws)
  where
    ws = words line

main :: IO ()
main = do readFile "day07.dat" >>= print . solution
